src/broadcast/broadcast.rs

```diff
@@ -26,16 +26,30 @@ pub struct Broadcast<N> {
     proposer_id: N,
     /// The Reed-Solomon erasure coding configuration.
     coding: Coding,
+    /// Original value message received from proposer.
+    value_message: Proof<Vec<u8>>,
     /// If we are the proposer: whether we have already sent the `Value` messages with the shards.
     value_sent: bool,
-    /// Whether we have already multicast `Echo`.
-    echo_sent: bool,
+    /// Whether we have already multicast `Echo` to our left nodes.
+    echo_sent_partial: bool,
```

Comment:
Yes that is correct. Isn't `echo_sent` in the original code also `true` iff `echos.contains_key(self.our_id()) == true`?

Reply:
Yes, I think you're right. Anyway, I'm always in favor of keeping redundancy out of data structures, to avoid all kinds of bugs where equivalent fields go out of sync…